domain/schema/model.go

```diff
@@ -558,11 +561,87 @@ INSERT INTO storage_kind VALUES
+    -- been fixed (since first implemented). We don't envisage
+    -- any change to how these are modelled.
+    --
+    -- Note: one might wonder why storage_pool below is not a
+    -- FK to a row defined in the storage pool table. This value
+    -- can also be one of the pool types. As with the comment on the
+    -- type column in the storage pool table, it's problematic to use a lookup
+    -- with an ID. Storage pools, once created, cannot be renamed so
+    -- this will not be able to become "orphaned".
+    storage_pool TEXT NOT NULL,
+    size         INT  NOT NULL,
+    count        INT  NOT NULL,
+    CONSTRAINT       fk_application_storage_constraint_application
+        FOREIGN KEY  (application_uuid)
+        REFERENCES   application(uuid),
+    PRIMARY KEY (application_uuid, storage_name)
+);
+
+-- Note that this is not unique; it speeds access by application.
+CREATE INDEX idx_application_storage_constraint
+ON application_storage_constraint (application_uuid);
+
+-- This table stores storage constraint values for each named storage item
+-- defined by the unit's current charm. If the charm is updated, then
+-- so too will be the rows in this table to reflect the current charm's
+-- storage definitions.
+-- Note: usually we just get the storage constraints off the application
+-- but need to allow for a unit's charm to temporarily diverge from that
+-- of its application.
+CREATE TABLE unit_storage_constraint (
+    unit_uuid               TEXT NOT NULL,
+    storage_name            TEXT NOT NULL,
+    -- These attributes are filled in by sourcing data from:
+    -- user supplied, model config, charm config, opinionated fallbacks.
+    -- By the time the row is written, all values are known.
+    -- Constraint value types (pool, size, count) hitherto have
+    -- been fixed (since first implemented). We don't envisage
+    -- any change to how these are modelled.
+    --
+    -- Note: one might wonder why storage_pool below is not a
+    -- FK to a row defined in the storage pool table. This value
+    -- can also be one of the pool types. As with the comment on the
+    -- type column in the storage pool table, it's problematic to use a lookup
+    -- with an ID. Storage pools, once created, cannot be renamed so
+    -- this will not be able to become "orphaned".
+    storage_pool TEXT NOT NULL,
+    size         INT  NOT NULL,
+    count        INT  NOT NULL,
+    CONSTRAINT       fk_unit_storage_constraint_unit
+        FOREIGN KEY  (unit_uuid)
+        REFERENCES   unit(uuid),
+    PRIMARY KEY (unit_uuid, storage_name)
+);
+
+-- Note that this is not unique; it speeds access by unit.
+CREATE INDEX idx_unit_storage_constraint
+ON unit_storage_constraint (unit_uuid);
+
 CREATE TABLE storage_instance (
     uuid            TEXT PRIMARY KEY,
     storage_kind_id INT NOT NULL,
     name            TEXT NOT NULL,
     life_id         INT NOT NULL,
+    -- Note: one might wonder why storage_pool below is not a
```

Comment:
Then we should have nullable columns for type ID and pool ID, with the appropriate FKs.

Reply:
As discussed live, this follows the same paradigm as used for storage pool type, reasoning is included in inline comments.